mentii/mentii
# Join public class

Changes to allow users to join public classes listed in the public class list. The following changes were made

1. app.py now has a joinClass method at the '/user/classes/' POST endpoint
2. class_ctrl.py has some added logging
3. In user_ctrl.py, users are now created without classCode lists. This is now added as a set the first time a user joins a class.
4. user_ctrl.py now has a joinClass  method. When called the class's code is added to the user's classCode set and the user's email is added to the class's student set .
5. Backend testing now uses mock to avoid trying to send an email, but we can use this for alot of other things as well
6. Added a test method fro the new prepForJsonDump method in ControllerResponse
7. Modified test_user_ctrl.py to setup both user and class tables in the same way that  test_class_ctrl.py does
8. Added a test method for user_ctrl.py's joinClass  method
9. ResponseCreation.py now has a prepForJsonDump method, which is called every time a response's payload is added to. This method confirms that the payload is ready to be converted to json by recursively replacing all sets and tuples with lists.
10. browse.component.ts can now subscribe to and handle the user service call to joinClass
11. Join class button now actually calls methods to join class
12. user.service.ts now has a joinClass method to call the backend's '/user/classes/' POST endpoint

To test: 
register and activate a new student. If there is already a public class listed, click the join class button to join. It should be removed from the public class list and added to the student's class list.  If there is not a public class use the admin account to add a new class, then switch back to the student and join the class.
